core/state/caching_versa_db.go

```diff
@@ -0,0 +1,297 @@
+	}
+	if cv.root.Cmp(root) != 0 {
+		panic("account root mismatch, on open storage tree")
+	}
+
+	version, account, err := cv.accTree.getAccountWithVersion(address)
+	if err != nil {
+		return nil, err
+	}
+	if account.Root.Cmp(stateRoot) != 0 {
+		return nil, fmt.Errorf("state root mismatch")
+	}
+
+	handler, err := cv.versionDB.OpenTree(cv.state, version, crypto.Keccak256Hash(address.Bytes()), stateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	tree := &VersaTree{
+		db:        cv.versionDB,
+		handler:   handler,
+		version:   version,
+		root:      root,
+		stateRoot: stateRoot,
+		address:   address,
+	}
+	return tree, nil
+}
+
+// Flush unique to versa
+func (cv *cachingVersaDB) Flush() error {
+	return cv.versionDB.Flush(cv.state)
+}
+
+// Release unique to versa
+func (cv *cachingVersaDB) Release() error {
+	if err := cv.versionDB.CloseState(cv.state); err != nil {
+		return nil
+	}
+	cv.release()
+	return nil
+}
+
+func (cv *cachingVersaDB) release() {
+	cv.hasState.Store(false)
+	cv.accTree = nil
+}
+
+func (cv *cachingVersaDB) CopyTrie(Trie) Trie {
+	//TODO:: support in the future
+	return nil
+}
+
+func (cv *cachingVersaDB) ContractCode(addr common.Address, codeHash common.Hash) ([]byte, error) {
+	code, _ := cv.codeCache.Get(codeHash)
+	if len(code) > 0 {
+		return code, nil
+	}
+	code = rawdb.ReadCodeWithPrefix(cv.codeDB, codeHash)
+	if len(code) > 0 {
+		cv.codeCache.Add(codeHash, code)
+		cv.codeSizeCache.Add(codeHash, len(code))
+		return code, nil
+	}
+	return nil, errors.New("not found")
+}
+
+func (cv *cachingVersaDB) ContractCodeSize(addr common.Address, codeHash common.Hash) (int, error) {
+	if cached, ok := cv.codeSizeCache.Get(codeHash); ok {
+		return cached, nil
+	}
+	code, err := cv.ContractCode(addr, codeHash)
+	return len(code), err
+}
+
+func (cv *cachingVersaDB) ContractCodeWithPrefix(address common.Address, codeHash common.Hash) ([]byte, error) {
+	code, _ := cv.codeCache.Get(codeHash)
+	if len(code) > 0 {
+		return code, nil
+	}
+	code = rawdb.ReadCodeWithPrefix(cv.codeDB, codeHash)
+	if len(code) > 0 {
+		cv.codeCache.Add(codeHash, code)
+		cv.codeSizeCache.Add(codeHash, len(code))
+		return code, nil
+	}
+	return nil, errors.New("not found")
+}
+
+func (cv *cachingVersaDB) DiskDB() ethdb.KeyValueStore {
+	return cv.codeDB
+}
+
+func (cv *cachingVersaDB) TrieDB() *triedb.Database {
+	return cv.triedb
+}
+
+func (cv *cachingVersaDB) NoTries() bool {
+	// TODO:: not support fastnode
+	return false
+}
+
+type VersaTree struct {
+	db          versa.Database
+	handler     versa.TreeHandler
+	version     int64
+	accountTree bool
+
+	// TODO:: debugging, used for logging
+	root      common.Hash
+	stateRoot common.Hash
+	address   common.Address
+}
+
+func (vt *VersaTree) GetKey(key []byte) []byte {
+	_, val, err := vt.db.Get(vt.handler, key)
+	if err != nil {
+		log.Warn("failed to get key from version db")
+	}
+	return val
+}
+
+func (vt *VersaTree) GetAccount(address common.Address) (*types.StateAccount, error) {
```

Comment:
I think the version of the account should be returned to the caller, and maintained by the account itself, so that it can be provided to versaDB when opening the storage tree.

But if do this, a lot of changes will be required. Therefore, we need to do some trade off whether the version should be known to the caller.

Reply:
Similarly, currently the upper layer does not have a strong dependency on the version, so it can be transparent to the upper layer for the time being. If needed in the future, it can be exposed then. The focus at this stage should not be on this.